src/pudl/transform/eia861.py

```diff
@@ -900,330 +1056,143 @@ def balancing_authority(tfr_dfs):
-    )
-    # We merge on ba_code then drop it, b/c only BA ID exists in all years consistently:
-    late_date_ba_util_state = (
-        late_date_ba_code_util_state.merge(late_ba_code_id, how="outer")
-        .drop("balancing_authority_code_eia", axis="columns")
-        .drop_duplicates()
+    ###########################################################################
+    # Tidy Data:
+    ###########################################################################
+    logger.info("Tidying the EIA 861 Sales table.")
+    tidy_sales, idx_cols = _tidy_class_dfs(
+        raw_sales,
+        df_name="Sales",
+        idx_cols=idx_cols,
+        class_list=CUSTOMER_CLASSES,
+        class_type="customer_class",
     )
 
-    tfr_dfs["balancing_authority_assn_eia861"] = (
-        pd.concat([early_date_ba_util_state, late_date_ba_util_state])
-        .dropna(
-            subset=[
-                "balancing_authority_id_eia",
-            ]
-        )
-        .pipe(apply_pudl_dtypes, group="eia")
+    # remove duplicates on the primary key columns + customer_class -- there
+    # are lots of records that have reporting errors in the form of duplicates.
+    # Many of them include different values and are therefore impossible to tell
+    # which are "correct". The following function drops all but the first of
+    # these duplicate entries.
+    deduped_sales = _drop_dupes(df=tidy_sales, df_name="Sales", subset=idx_cols)
+
+    ###########################################################################
+    # Transform Values:
+    # * Turn 1000s of dollars back into dollars
+    # * Re-code data_observed to boolean:
+    #   * O="observed" => True
+    #   * I="imputed" => False
+    # * Change the form code (A, B, C, D) into the business model that it
+    #   corresponds to (retail vs. energy_services), which in combination with
+    #   the service_type column (energy, delivery, bundled) will now serve as
+    #   part of the primary key for the table.
+    ###########################################################################
+    logger.info("Performing value transformations on EIA 861 Sales table.")
+    transformed_sales = deduped_sales.assign(
+        sales_revenue=lambda x: _thousand_to_one(x.sales_revenue),
+        data_observed=lambda x: x.data_observed.replace(
+            {
+                "O": True,
+                "I": False,
+            }
+        ),
+        business_model=lambda x: x.business_model.replace(
+            {
+                "A": "retail",
+                "B": "retail",
+                "C": "retail",
+                "D": "energy_services",
+            }
+        ),
+        service_type=lambda x: x.service_type.str.lower().replace(
+            {"bundle": "bundled"}
+        ),
+        short_form=lambda x: _make_yn_bool(x.short_form),
     )
-    return tfr_dfs
 
+    return _post_process(transformed_sales)
```

Comment:
Excited to apply your FERC transform classes to our EIA pipelines. Seems like a perfect fit.

Reply:
I think the EIA transforms will probably be simpler than the FERC ones too. Not sure when we're going to find the time to refactor them though.

Reply:
Yeah seems like a lower-priority thing. Maybe a decent project for an excited external contributor.